obelisk/cpp/obelisk_cpp/include/obelisk_sim_robot.h

```diff
@@ -0,0 +1,139 @@
+
+        /**
+         * @brief cleans up the node.
+         *
+         * @param prev_state the state of the ros node.
+         */
+        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn virtual on_cleanup(
+            const rclcpp_lifecycle::State& prev_state) {
+            this->ObeliskRobot<ControlMessageT>::on_cleanup(prev_state);
+
+            // Cleanup the sim thread
+            bool current_thread_status = stop_thread_;
+            stop_thread_               = true;
+            if (sim_thread_.joinable()) {
+                sim_thread_.join();
+            }
+
+            if (!current_thread_status) {
+                RCLCPP_INFO_STREAM(this->get_logger(), "Simulation thread stopped.");
+            }
+
+            return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+        }
+
+        /**
+         * @brief shutsdown the node the node.
+         *
+         * @param prev_state the state of the ros node.
+         */
+        rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn virtual on_shutdown(
+            const rclcpp_lifecycle::State& prev_state) {
+            this->ObeliskRobot<ControlMessageT>::on_shutdown(prev_state);
+
+            return on_cleanup(prev_state);
+        }
+
+      protected:
+        /**
+         * @brief Publish the TrueSimState of the simulator.
+         *
+         * This is the timer callback that publishes the TrueSimState and is expected to call
+         * self.publisher_true_sim_state internally. Note that the true sim state message is still returned afterwards,
+         * mostly for logging/debugging purposes. The publish call is the important part, NOT the returned value, since
+         * the topic is what the ObeliskEstimator subscribes to.
+         *
+         * @return obelisk_true_sim_state_msg: An Obelisk message type containing the true sim state.
+         */
+        virtual TrueSimState PublishTrueSimState() = 0;
+
+        /**
+         * @brief Run the simulator
+         *
+         * The control input into the simulator is accessed through the shared control array. If any simulator
+         * configuration needs to occur, then the on_configure function should be overridden. This function should run
+         * the simulator loop and update the state of the simulator as long as the node is active.
+         *
+         * This is where the simulation loop is run. The simulation loop should always be check the stop_thread_ flag to
+         * determine when to stop.
+         */
+        virtual void RunSimulator() = 0;
+
+        void StartSim() { this->RunSimulator(); }
+
+        // ---------- Member Variables --------- //
+        // Hold the thread where the simulation takes place
+        std::thread sim_thread_;
+
+        // Flag to signal stopping the thread
+        std::atomic<bool> stop_thread_;
+
+        std::string state_timer_key_;
+        std::string state_pub_key_;
```

Comment:
On the python side i ended up moving this logic to the mujoco-specific sim class, mostly because the multiprocessing stuff felt a bit too specific to always apply to the generic sim robot API. We should discuss where this should go.

Reply:
I moved the shared data to the mujoco specific class, but I assumed that having the simulation in a separate thread, seemed generic enough that every simulation interface would want it. I can't think of sim that you wouldn't want to do this for, so I think we should leave the multi-threading framework here.